vllm/entrypoints/openai/api_server.py

```diff
@@ -106,6 +108,21 @@ async def build_async_engine_client(args) -> AsyncIterator[AsyncEngineClient]:
 
     # Otherwise, use the multiprocessing AsyncLLMEngine.
     else:
+        if "PROMETHEUS_MULTIPROC_DIR" not in os.environ:
+            # Make TemporaryDirectory for prometheus multiprocessing
+            # Note: global TemporaryDirectory will be automatically
+            #   cleaned up upon exit.
+            global prometheus_multiproc_dir
+            prometheus_multiproc_dir = tempfile.TemporaryDirectory()
+            os.environ[
+                "PROMETHEUS_MULTIPROC_DIR"] = prometheus_multiproc_dir.name
+        else:
+            logger.warning(
+                "Found PROMETHEUS_MULTIPROC_DIR was set by user. "
+                "This directory must be wiped between vLLM runs or "
+                "you will find inaccurate metrics. Unset the variable "
+                "and vLLM will properly handle cleanup.")
```

Comment:
You should actually be setting up a shutdown hook as mentioned in point 3 in http://prometheus.github.io/client_python/multiprocess/. Processes will live and die and leave their individual metrics files laying around. And they will be collected and reported with processes that are still living. Which results in incorrect statistics.

Reply:
Per my other comment, processes won't be coming and going and we'll clean up the temp dir on shutdown, so this shouldn't be an issue.

Reply:
It’s not a bad idea to add this. I’ll do so after the process join